### PR TITLE
Make parseInt unicode tolerant

### LINKED ISSUE #405: Unicode support in implementation of parseInt
Hello.

Implementation of `parseInt` (ecma-builtin-global.cpp) should be updated to use UTF-8 iterators for processing characters buffer.

Comment of @zherczeg in #394:
> Is this correct? If I understand correctly, we just pass a byte, rather than an uint16_t character. I think we should rewrite this to use the UTF iterators, since there are line terminators and white spaces outside of the [0-127] byte range.

### PR DESCRIPTION
**Pull request by galpeter**
Fixes #405 

JerryScript-DCO-1.0-Signed-off-by: Peter Gal pgal.u-szeged@partner.samsung.com

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 1]
```diff
@@ -185,47 +195,55 @@ ecma_builtin_global_object_parse_int (ecma_value_t this_arg __attr_unused___, /*
...
       }
 
       /* 11. Check if characters are in [0, Radix - 1]. We also convert them to number values in the process. */
-      for (lit_utf8_size_t i = start; i < end; i++)
+      lit_utf8_iterator_seek (&iter, start);
+      while (!lit_utf8_iterator_is_eos (&iter))
       {
-        if ((utf8_string_buff[i]) >= LIT_CHAR_LOWERCASE_A && utf8_string_buff[i] <= LIT_CHAR_LOWERCASE_Z)
+        ecma_char_t current_char = lit_utf8_iterator_read_next (&iter);
+        ecma_number_t current_number;
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: I think we should use an integer type here instead of an `ecma_number_t`.

**Comment by zherczeg** [comment]
LGTM

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 2]
```diff
@@ -140,15 +148,17 @@ ecma_builtin_global_object_parse_int (ecma_value_t this_arg __attr_unused___, /*
...
       sign = -1;
     }
 
     /* 5. */
-    if (utf8_string_buff[start] == LIT_CHAR_MINUS || utf8_string_buff[start] == LIT_CHAR_PLUS)
+    if (current == LIT_CHAR_MINUS || current == LIT_CHAR_PLUS)
     {
-      start++;
+      start = lit_utf8_iterator_get_pos (&iter);
+      current = lit_utf8_iterator_read_next (&iter);
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: Shouldn't we check for `eos` before doing this?

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp (hunk 1 above)
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: good point, will fix it in the update.

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp (hunk 2 above)
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: Yeah, good catch!

**Comment by galpeter** [comment]
PR updated.

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp [hunk 3]
```diff
@@ -185,47 +198,55 @@ ecma_builtin_global_object_parse_int (ecma_value_t this_arg __attr_unused___, /*
...
       if (strip_prefix)
       {
-        if (end - start >= 2
-            && utf8_string_buff[start] == LIT_CHAR_0
-            && (utf8_string_buff[start + 1] == LIT_CHAR_LOWERCASE_X
-                || utf8_string_buff[start + 1] == LIT_CHAR_UPPERCASE_X))
+        if (end.offset - start.offset >= 2 && current == LIT_CHAR_0)
         {
-          start += 2;
+          ecma_char_t next = lit_utf8_iterator_peek_next (&iter);
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: There probably should be an `eos` check here as well.

Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: In this case it is not needed, as the offset check in the `if` statement above already checked that there are at least two characters, so we are not at the end of the string yet.

Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp]: Ah, okay.

**Comment by dbatyai** [comment]
lgtm

**Comment by sand1k** [comment]
LGTM.

**Comment by sand1k** [comment]
LGTM.

**Comment by egavrin** [comment]
`make push`

**Comment by dbatyai** [comment]
Landed: 3fbe54393f4900ea03b452326d1e96dbd9bf2d3a

**Pull request closed without merging**